File: Java/Hussain Sir Java/Fundamentals/Platform/Concurrency/ThreadCoordTest.java
class ThreadCoordTest{

	private volatile static int value = 0;
	private static Object coordinator = new Object();

	private static void consume(){
		System.out.printf("Consumer<%x> ready...%n", Thread.currentThread().hashCode());
		//while(value == 0) Thread.yield();
		synchronized(coordinator){
			try{
				coordinator.wait();
			}catch(InterruptedException e){}
			System.out.printf("Processed value = %d%n", value * value);
		}
	}

	private static void produce(){
		System.out.printf("Producer<%x> ready...%n", Thread.currentThread().hashCode());
		int result = Worker.doWork();
		synchronized(coordinator){
			value = result;
			coordinator.notify();
		}
	}

	public static void main(String[] args) throws Exception{
		Thread producer = new Thread(ThreadCoordTest::produce);
		producer.start();
		Thread consumer = new Thread(ThreadCoordTest::consume);
		consumer.start();
	}
}

File: Fundamentals/Platform/Concurrency/threadcoordtest.cs
using System;
using System.Threading;

static class Program
{
	private static volatile int value = 0;
	private static object coordinator = new object();
	
	private static void Consume()
	{
		Console.WriteLine("Consumer<{0}> ready...", Thread.CurrentThread.ManagedThreadId);

		//while(value == 0) Thread.Yield();
		lock(coordinator)
		{
			Monitor.Wait(coordinator);
			Console.WriteLine("Processed value = {0}", value * value);
		}
	}

	private static void Produce()
	{
		Console.WriteLine("Producer<{0}> ready...", Thread.CurrentThread.ManagedThreadId);

		int result = Worker.DoWork();
		
		lock(coordinator)
		{
			value = result;
			Monitor.Pulse(coordinator);
		}

	}

	public static void Main()
	{
		var producer = new Thread(Produce);
		producer.Start();
	
		var consumer = new Thread(Consume);
		consumer.Start();
	}
}	